File: docs/Drasil/src/cpp/Calculations.cpp
#include "Calculations.hpp"

#include <fstream>
#include <iostream>
#include <math.h>
#include <string>

#include "InputParameters.hpp"

using std::ofstream;
using std::string;

double func_F_vect_AC(InputParameters &inParams) {
    ofstream outfile;
    outfile.open("log.txt", std::fstream::app);
    outfile << "function func_F_vect_AC called with inputs: {" << std::endl;
    outfile << "  inParams = ";
    outfile << "Instance of InputParameters object" << std::endl;
    outfile << "  }" << std::endl;
    outfile.close();
    
    return -inParams.F_vect_Ay / sin(inParams.θ_1);
}

double func_F_vect_BC(InputParameters &inParams) {
    ofstream outfile;
    outfile.open("log.txt", std::fstream::app);
    outfile << "function func_F_vect_BC called with inputs: {" << std::endl;
    outfile << "  inParams = ";
    outfile << "Instance of InputParameters object" << std::endl;
    outfile << "  }" << std::endl;
    outfile.close();
    
    return -inParams.F_vect_By / sin(inParams.θ_2);
}

double func_F_vect_AD(InputParameters &inParams, double F_vect_AC) {
    ofstream outfile;
    outfile.open("log.txt", std::fstream::app);
    outfile << "function func_F_vect_AD called with inputs: {" << std::endl;
    outfile << "  inParams = ";
    outfile << "Instance of InputParameters object";
    outfile << ", " << std::endl;
    outfile << "  F_vect_AC = ";
    outfile << F_vect_AC << std::endl;
    outfile << "  }" << std::endl;
    outfile.close();
    
    return -F_vect_AC / cos(inParams.θ_1);
}

double func_F_vect_BD(InputParameters &inParams, double F_vect_BC) {
    ofstream outfile;
    outfile.open("log.txt", std::fstream::app);
    outfile << "function func_F_vect_BD called with inputs: {" << std::endl;
    outfile << "  inParams = ";
    outfile << "Instance of InputParameters object";
    outfile << ", " << std::endl;
    outfile << "  F_vect_BC = ";
    outfile << F_vect_BC << std::endl;
    outfile << "  }" << std::endl;
    outfile.close();
    
    return -F_vect_BC / cos(inParams.θ_2);
}

File: docs/Drasil/src/swift/Calculations.swift
/** Calculations.swift
    Provides functions for calculating the outputs
    - Authors: Ting-Yu Wu
*/
import Foundation

/** Calculates force between joint A and C (N)
    - Parameter inParams: structure holding the input values
    - Returns: force between joint A and C (N)
*/
func func_F_vect_AC(_ inParams: inout InputParameters) throws -> Double {
    var outfile: FileHandle
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("function func_F_vect_AC called with inputs: {".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  inParams = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("Instance of InputParameters object".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  }".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
    
    return -inParams.F_vect_Ay / sin(inParams.θ_1)
}

/** Calculates force between joint B and C (N)
    - Parameter inParams: structure holding the input values
    - Returns: force between joint B and C (N)
*/
func func_F_vect_BC(_ inParams: inout InputParameters) throws -> Double {
    var outfile: FileHandle
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("function func_F_vect_BC called with inputs: {".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  inParams = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("Instance of InputParameters object".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  }".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
    
    return -inParams.F_vect_By / sin(inParams.θ_2)
}

/** Calculates force between joint A and D (N)
    - Parameter inParams: structure holding the input values
    - Parameter F_vect_AC: force between joint A and C (N)
    - Returns: force between joint A and D (N)
*/
func func_F_vect_AD(_ inParams: inout InputParameters, _ F_vect_AC: Double) throws -> Double {
    var outfile: FileHandle
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("function func_F_vect_AD called with inputs: {".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  inParams = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("Instance of InputParameters object".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(", ".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  F_vect_AC = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(String(F_vect_AC).utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  }".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
    
    return -F_vect_AC / cos(inParams.θ_1)
}

/** Calculates force between joint B and D (N)
    - Parameter inParams: structure holding the input values
    - Parameter F_vect_BC: force between joint B and C (N)
    - Returns: force between joint B and D (N)
*/
func func_F_vect_BD(_ inParams: inout InputParameters, _ F_vect_BC: Double) throws -> Double {
    var outfile: FileHandle
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("function func_F_vect_BD called with inputs: {".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  inParams = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("Instance of InputParameters object".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(", ".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  F_vect_BC = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(String(F_vect_BC).utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  }".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
    
    return -F_vect_BC / cos(inParams.θ_2)
}
